flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java

```diff
@@ -0,0 +1,320 @@
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+import org.apache.flink.util.function.BiConsumerWithException;
+import org.apache.flink.util.function.RunnableWithException;
+
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+import static org.apache.flink.runtime.state.ChannelPersistenceITCase.getStreamFactoryFactory;
+import static org.apache.flink.util.ExceptionUtils.findThrowable;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * {@link ChannelStateWriterImpl} lifecycle tests.
+ */
+public class ChannelStateWriterImplTest {
+	private static final long CHECKPOINT_ID = 42L;
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testAddEventBuffer() {
+		NetworkBuffer dataBuf = getBuffer();
+		NetworkBuffer eventBuf = getBuffer();
+		eventBuf.tagAsEvent();
+		ChannelStateWriterImpl writer = openWriter();
+		callStart(writer);
+		try {
+			writer.addInputData(CHECKPOINT_ID, new InputChannelInfo(1, 1), 1, eventBuf, dataBuf);
+		} finally {
+			assertTrue(dataBuf.isRecycled());
+		}
+	}
+
+	@Test
+	public void testResultCompletion() throws IOException {
+		ChannelStateWriteResult result;
+		try (ChannelStateWriterImpl writer = openWriter()) {
+			callStart(writer);
+			result = writer.getWriteResult(CHECKPOINT_ID);
+			assertFalse(result.resultSubpartitionStateHandles.isDone());
+			assertFalse(result.inputChannelStateHandles.isDone());
+		}
+		assertTrue(result.inputChannelStateHandles.isDone());
+		assertTrue(result.resultSubpartitionStateHandles.isDone());
+	}
+
+	@Test
+	public void testAbort() throws Exception {
+		NetworkBuffer buffer = getBuffer();
+		runWithSyncWorker((writer, worker) -> {
+			callStart(writer);
+			callAddInputData(writer, buffer);
+			callAbort(writer);
+			worker.processAllRequests();
+			assertTrue(writer.getWriteResult(CHECKPOINT_ID).isDone());
+			assertTrue(buffer.isRecycled());
+		});
+	}
+
+	@Test
+	public void testAbortIgnoresMissing() throws Exception {
+		runWithSyncWorker(this::callAbort);
+	}
+
+	@Test(expected = TestException.class)
+	public void testBuffersRecycledOnError() throws Exception {
+		unwrappingError(TestException.class, () -> {
+			NetworkBuffer buffer = getBuffer();
+			try (ChannelStateWriterImpl writer = new ChannelStateWriterImpl(new ConcurrentHashMap<>(), failingWorker(), 5)) {
+				writer.open();
+				callAddInputData(writer, buffer);
+			} finally {
+				assertTrue(buffer.isRecycled());
+			}
+		});
+	}
+
+	@Test
+	public void testBuffersRecycledOnClose() throws IOException {
+		NetworkBuffer buffer = getBuffer();
+		try (ChannelStateWriterImpl writer = openWriter()) {
+			callStart(writer);
+			callAddInputData(writer, buffer);
+			assertFalse(buffer.isRecycled());
```

Comment:
seems unstable results? After `callAddInputData`, if the enqueued buffer is already dispatched to be executed by internal thread, then the buffer should be recycled when this assert calls. Although this probability is very small, but in theory it is not stable and easily fragile.

Reply:
You're right, fixed it.